*   Name: Priyanka Mulagandla
*   Net ID: pm3392
*   N18226862

**Topic:** Popular online Games

**Data source:** https://api.rawg.io/docs/#operation/games_suggested_read

**Target Audience:** Gamers, App developers, App investors, App Publishers

**Data analysis questions**
1. What are the top 10 popular game genres?
2. How many games are releasd through years?
3. What are the top 8 Gaming Platforms with highest number of game releases?
4. What are the top 10 rated games?
5. What are the top 10 downloaded games?

#d1: Library imports


In [ ]:
#d1: Library imports
!pip install PyMySQL
import requests
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly
import plotly.express as px
import plotly.graph_objects as go
from plotly.offline import init_notebook_mode,iplot
import pymysql.cursors
import pymysql

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 43 kB 1.4 MB/s 


#d2: Data pre-processing


In [ ]:
#Fetch the data from API endpoint
#page has 40 rows, so used for loop to iterate and get 1000 rows
df1 = pd.DataFrame()
for i in range(25):
  url = 'https://api.rawg.io/api/games?key=4bc2ebc60298479ebf7f6cfa40adde63&page='+str(i+1)+'&page_size=40'
  df = pd.DataFrame(requests.get(url).json()['results'])
  df1 = df1.append(df, ignore_index = True)
df1

id                             slug  \
0      3498               grand-theft-auto-v   
1      3328          the-witcher-3-wild-hunt   
2      4200                         portal-2   
3      5286                      tomb-raider   
4      4291  counter-strike-global-offensive   
..      ...                              ...   
995   52330                      the-council   
996  455532              star-wars-squadrons   
997   28627                           halo-4   
998     703                       let-it-die   
999   11407                         banished   

                                 name    released    tba  \
0                  Grand Theft Auto V  2013-09-17  False   
1            The Witcher 3: Wild Hunt  2015-05-18  False   
2                            Portal 2  2011-04-18  False   
3                  Tomb Raider (2013)  2013-03-05  False   
4    Counter-Strike: Global Offensive  2012-08-21  False   
..                                ...         ...    ...   
995                       The Council  2018-03-12  False   
996              Star Wars: Squadrons  2020-10-02  False   
997                            Halo 4  2012-11-06  False   
998                        LET IT DIE  2016-12-03  False   
999                          Banished  2014-02-17  False   

                                      background_image  rating  rating_top  \
0    https://media.rawg.io/media/games/456/456dea5e...    4.47           5   
1    https://media.rawg.io/media/games/618/618c2031...    4.67           5   
2    https://media.rawg.io/media/games/328/3283617c...    4.62           5   
3    https://media.rawg.io/media/games/021/021c4e21...    4.05           4   
4    https://media.rawg.io/media/games/736/73619bd3...    3.57           4   
..                                                 ...     ...         ...   
995  https://media.rawg.io/media/games/123/123e0357...    3.63           4   
996  https://media.rawg.io/media/games/b19/b191ff2b...    3.23           4   
997  https://media.rawg.io/media/games/efd/efdc57a6...    3.80           4   
998  https://media.rawg.io/media/games/d4a/d4a85e0d...    2.97           3   
999  https://media.rawg.io/media/screenshots/0bf/0b...    3.85           4   

                                               ratings  ratings_count  ...  \
0    [{'id': 5, 'title': 'exceptional', 'count': 35...           6012  ...   
1    [{'id': 5, 'title': 'exceptional', 'count': 44...           5653  ...   
2    [{'id': 5, 'title': 'exceptional', 'count': 35...           4984  ...   
3    [{'id': 4, 'title': 'recommended', 'count': 21...           3498  ...   
4    [{'id': 4, 'title': 'recommended', 'count': 14...           3050  ...   
..                                                 ...            ...  ...   
995  [{'id': 4, 'title': 'recommended', 'count': 16...            287  ...   
996  [{'id': 4, 'title': 'recommended', 'count': 97...            204  ...   
997  [{'id': 4, 'title': 'recommended', 'count': 26...            520  ...   
998  [{'id': 3, 'title': 'meh', 'count': 143, 'perc...            313  ...   
999  [{'id': 4, 'title': 'recommended', 'count': 16...            276  ...   

     saturated_color  dominant_color  \
0             0f0f0f          0f0f0f   
1             0f0f0f          0f0f0f   
2             0f0f0f          0f0f0f   
3             0f0f0f          0f0f0f   
4             0f0f0f          0f0f0f   
..               ...             ...   
995           0f0f0f          0f0f0f   
996           0f0f0f          0f0f0f   
997           0f0f0f          0f0f0f   
998           0f0f0f          0f0f0f   
999           0f0f0f          0f0f0f   

                                             platforms  \
0    [{'platform': {'id': 4, 'name': 'PC', 'slug': ...   
1    [{'platform': {'id': 186, 'name': 'Xbox Series...   
2    [{'platform': {'id': 14, 'name': 'Xbox 360', '...   
3    [{'platform': {'id': 18, 'name': 'PlayStation ...   
4    [{'platform': {'id': 4, 'name': 'PC', 'slug': ...   
..      

In [ ]:

#created a list to store values in the column 'parent_platforms'
for index, row in df1.iterrows():
    listt = row['parent_platforms']
    listt2 = []
    for element in listt:
      if element['platform']['name'] != None :
        listt2.append(element['platform']['name'])
    df1['parent_platforms'].iloc[index] = listt2
df1

/usr/local/lib/python3.8/dist-packages/pandas/core/indexing.py:1732: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



id                             slug  \
0      3498               grand-theft-auto-v   
1      3328          the-witcher-3-wild-hunt   
2      4200                         portal-2   
3      5286                      tomb-raider   
4      4291  counter-strike-global-offensive   
..      ...                              ...   
995   52330                      the-council   
996  455532              star-wars-squadrons   
997   28627                           halo-4   
998     703                       let-it-die   
999   11407                         banished   

                                 name    released    tba  \
0                  Grand Theft Auto V  2013-09-17  False   
1            The Witcher 3: Wild Hunt  2015-05-18  False   
2                            Portal 2  2011-04-18  False   
3                  Tomb Raider (2013)  2013-03-05  False   
4    Counter-Strike: Global Offensive  2012-08-21  False   
..                                ...         ...    ...   
995                       The Council  2018-03-12  False   
996              Star Wars: Squadrons  2020-10-02  False   
997                            Halo 4  2012-11-06  False   
998                        LET IT DIE  2016-12-03  False   
999                          Banished  2014-02-17  False   

                                      background_image  rating  rating_top  \
0    https://media.rawg.io/media/games/456/456dea5e...    4.47           5   
1    https://media.rawg.io/media/games/618/618c2031...    4.67           5   
2    https://media.rawg.io/media/games/328/3283617c...    4.62           5   
3    https://media.rawg.io/media/games/021/021c4e21...    4.05           4   
4    https://media.rawg.io/media/games/736/73619bd3...    3.57           4   
..                                                 ...     ...         ...   
995  https://media.rawg.io/media/games/123/123e0357...    3.63           4   
996  https://media.rawg.io/media/games/b19/b191ff2b...    3.23           4   
997  https://media.rawg.io/media/games/efd/efdc57a6...    3.80           4   
998  https://media.rawg.io/media/games/d4a/d4a85e0d...    2.97           3   
999  https://media.rawg.io/media/screenshots/0bf/0b...    3.85           4   

                                               ratings  ratings_count  ...  \
0    [{'id': 5, 'title': 'exceptional', 'count': 35...           6012  ...   
1    [{'id': 5, 'title': 'exceptional', 'count': 44...           5653  ...   
2    [{'id': 5, 'title': 'exceptional', 'count': 35...           4984  ...   
3    [{'id': 4, 'title': 'recommended', 'count': 21...           3498  ...   
4    [{'id': 4, 'title': 'recommended', 'count': 14...           3050  ...   
..                                                 ...            ...  ...   
995  [{'id': 4, 'title': 'recommended', 'count': 16...            287  ...   
996  [{'id': 4, 'title': 'recommended', 'count': 97...            204  ...   
997  [{'id': 4, 'title': 'recommended', 'count': 26...            520  ...   
998  [{'id': 3, 'title': 'meh', 'count': 143, 'perc...            313  ...   
999  [{'id': 4, 'title': 'recommended', 'count': 16...            276  ...   

     saturated_color  dominant_color  \
0             0f0f0f          0f0f0f   
1             0f0f0f          0f0f0f   
2             0f0f0f          0f0f0f   
3             0f0f0f          0f0f0f   
4             0f0f0f          0f0f0f   
..               ...             ...   
995           0f0f0f          0f0f0f   
996           0f0f0f          0f0f0f   
997           0f0f0f          0f0f0f   
998           0f0f0f          0f0f0f   
999           0f0f0f          0f0f0f   

                                             platforms  \
0    [{'platform': {'id': 4, 'name': 'PC', 'slug': ...   
1    [{'platform': {'id': 186, 'name': 'Xbox Series...   
2    [{'platform': {'id': 14, 'name': 'Xbox 360', '...   
3    [{'platform': {'id': 18, 'name': 'PlayStation ...   
4    [{'platform': {'id': 4, 'name': 'PC', 'slug': ...   
..      

In [ ]:
#checking for null values
df1.isna().sum()

id                       0
slug                     0
name                     0
released                 0
tba                      0
background_image         0
rating                   0
rating_top               0
ratings                  0
ratings_count            0
reviews_text_count       0
added                    0
added_by_status          0
metacritic             128
playtime                 0
suggestions_count        0
updated                  0
user_game             1000
reviews_count            0
saturated_color          0
dominant_color           0
platforms                0
parent_platforms         0
genres                   0
stores                   0
clip                  1000
tags                     0
esrb_rating            279
short_screenshots        0
dtype: int64

In [ ]:
#drop invalid columns
df1.drop(['slug','rating_top'], axis=1, inplace = True)


In [ ]:
df1.drop(['tba', 'background_image', 'reviews_text_count','metacritic', 'clip' ,'short_screenshots', 'platforms', 'saturated_color' ,'dominant_color'],axis =1, inplace = True)

In [ ]:
df1.drop(['ratings', 'added_by_status', 'suggestions_count', 'updated'],axis =1, inplace = True)

In [ ]:
df1.drop(['stores'],axis = 1, inplace = True)

In [ ]:
df1.drop(['tags','playtime','reviews_count'],axis = 1, inplace = True)

In [ ]:
df1.drop(['user_game'], axis = 1, inplace = True)

In [ ]:
#printing to check the columns present
df1.columns

Index(['id', 'name', 'released', 'rating', 'ratings_count', 'added',
       'parent_platforms', 'genres', 'esrb_rating'],
      dtype='object')

In [ ]:
#handling outliers
#picking genre from genres values in json 'genre' column
for index,row in df1.iterrows():
  if len(row['genres'])>0:
    df1['genres'].iloc[index] = row['genres'][0]['name']
  else:
    df1['genres'].iloc[index] = 'Other'
df1.head(5)

/usr/local/lib/python3.8/dist-packages/pandas/core/indexing.py:1732: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



id                              name    released  rating  ratings_count  \
0  3498                Grand Theft Auto V  2013-09-17    4.47           6012   
1  3328          The Witcher 3: Wild Hunt  2015-05-18    4.67           5653   
2  4200                          Portal 2  2011-04-18    4.62           4984   
3  5286                Tomb Raider (2013)  2013-03-05    4.05           3498   
4  4291  Counter-Strike: Global Offensive  2012-08-21    3.57           3050   

   added                                 parent_platforms   genres  \
0  18279                          [PC, PlayStation, Xbox]   Action   
1  17195                [PC, PlayStation, Xbox, Nintendo]   Action   
2  16180  [PC, PlayStation, Xbox, Apple Macintosh, Linux]  Shooter   
3  14286         [PC, PlayStation, Xbox, Apple Macintosh]   Action   
4  13938                          [PC, PlayStation, Xbox]   Action   

                                         esrb_rating  
0      {'id': 4, 'name': 'Mature', 'slug': 'mature'}  
1      {'id': 4, 'name': 'Mature', 'slug': 'mature'}  
2  {'id': 2, 'name': 'Everyone 10+', 'slug': 'eve...  
3      {'id': 4, 'name': 'Mature', 'slug': 'mature'}  
4      {'id': 4, 'name': 'Mature', 'slug': 'mature'}

In [ ]:
#handling outliers
#picking values from jsons in the 'esrb_rating' column
for index, row in df1.iterrows():
  if row['esrb_rating'] is not None :
    df1['esrb_rating'].iloc[index] = row['esrb_rating']['name']


/usr/local/lib/python3.8/dist-packages/pandas/core/indexing.py:1732: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [ ]:
#replacing Null values
df1['esrb_rating'] = df1['esrb_rating'].fillna('Everyone')

In [ ]:
df1 = df1.replace(['Rating Pending'], 'Everyone')

In [ ]:
#checking for values in 'esrb_rating' column
df1['esrb_rating'].value_counts()

Mature          362
Everyone        325
Teen            188
Everyone 10+     92
Adults Only      33
Name: esrb_rating, dtype: int64

In [ ]:
#checking for NULL values in all columns
df1.isna().sum()

id                  0
name                0
released            0
rating              0
ratings_count       0
added               0
parent_platforms    0
genres              0
esrb_rating         0
dtype: int64

In [ ]:
#renaming the columns
renaming_df = {
    'released': 'released_date',
    'added' : 'downloads',
    'parent_platforms':'platforms',
    'esrb_rating':'age_rating'
}
df1.rename(columns=renaming_df, inplace = True)
df1

id                              name released_date  rating  \
0      3498                Grand Theft Auto V    2013-09-17    4.47   
1      3328          The Witcher 3: Wild Hunt    2015-05-18    4.67   
2      4200                          Portal 2    2011-04-18    4.62   
3      5286                Tomb Raider (2013)    2013-03-05    4.05   
4      4291  Counter-Strike: Global Offensive    2012-08-21    3.57   
..      ...                               ...           ...     ...   
995   52330                       The Council    2018-03-12    3.63   
996  455532              Star Wars: Squadrons    2020-10-02    3.23   
997   28627                            Halo 4    2012-11-06    3.80   
998     703                        LET IT DIE    2016-12-03    2.97   
999   11407                          Banished    2014-02-17    3.85   

     ratings_count  downloads  \
0             6012      18279   
1             5653      17195   
2             4984      16180   
3             3498      14286   
4             3050      13938   
..             ...        ...   
995            287       2043   
996            204       2039   
997            520       2038   
998            313       2038   
999            276       2037   

                                           platforms    genres    age_rating  
0                            [PC, PlayStation, Xbox]    Action        Mature  
1                  [PC, PlayStation, Xbox, Nintendo]    Action        Mature  
2    [PC, PlayStation, Xbox, Apple Macintosh, Linux]   Shooter  Everyone 10+  
3           [PC, PlayStation, Xbox, Apple Macintosh]    Action        Mature  
4                            [PC, PlayStation, Xbox]    Action        Mature  
..                                               ...       ...           ...  
995                          [PC, PlayStation, Xbox]    Action        Mature  
996                          [PC, PlayStation, Xbox]    Action          Teen  
997                                       [PC, Xbox]    Action        Mature  
998                                [PC, PlayStation]    Action        Mature  
999                                             [PC]  Strategy          Teen  

[1000 rows x 9 columns]

In [ ]:
#writing the values in the column 'platforms' into a set to see unique values
platform_set = set()
for index,row in df1.iterrows():
  platform_list = row['platforms']
  if(len(platform_list) > 0):
    for x in platform_list:
      platform_set.add(x)
print(platform_set)

{'PC', 'Apple Macintosh', 'Linux', 'SEGA', 'Xbox', 'Nintendo', 'Commodore / Amiga', 'Neo Geo', 'Web', 'Android', 'PlayStation', 'iOS'}


In [ ]:
!pip install PyMySQL

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 43 kB 2.0 MB/s 


In [ ]:
#import the cursors classa AND the entire pymysql library
import pymysql.cursors
import pymysql

# Connect to the database
connection = pymysql.connect(host='35.236.250.241',
                             user='root',
                             password='Pinky@1319',
                             db='data_project',
                             charset='utf8mb4',
                             cursorclass=pymysql.cursors.DictCursor)

#create an cursor object. The cursor allows you to execute operations such as SQL statements.
cursor = connection.cursor()


#d3: Inserting the data

In [ ]:
#final code
#insert data into the 'platforms' table
for index, row in enumerate(platform_set):
    name = row
    pid = index
    print(pid, name)
    print("Inserting platforms " + str(index+1))
    query_template ="INSERT INTO platforms (name) VALUES(%s)"
    query_parameters = (name)
    cursor.execute(query_template, query_parameters)
    #connection.commit() 

0 Neo Geo
Inserting platforms 1
1 PC
Inserting platforms 2
2 Android
Inserting platforms 3
3 Nintendo
Inserting platforms 4
4 Web
Inserting platforms 5
5 Apple Macintosh
Inserting platforms 6
6 Commodore / Amiga
Inserting platforms 7
7 iOS
Inserting platforms 8
8 Xbox
Inserting platforms 9
9 PlayStation
Inserting platforms 10
10 SEGA
Inserting platforms 11
11 Linux
Inserting platforms 12


In [ ]:
sql = 'select * from platforms'
cursor.execute(sql)

12

In [ ]:
df1.columns

Index(['id', 'name', 'released_date', 'rating', 'ratings_count', 'downloads',
       'platforms', 'genres', 'age_rating'],
      dtype='object')

In [ ]:
#final code
#Insert data into the 'games' table
for index, row in df1.iterrows():
    id = int(row['id'])
    name = row['name']
    released_date = row['released_date']
    rating = float(row['rating'])
    ratings_count = int(row['ratings_count'])
    downloads = int(row['downloads'])
    genres = row['genres']
    age_rating = row['age_rating']
    print(id, name, released_date, rating, ratings_count, downloads, genres, age_rating)
    print("Inserting platforms " + str(index+1))
    query_template ="INSERT INTO games (id, name, released_date, rating, ratings_count, downloads, genres, age_rating) VALUES(%s,%s,%s,%s,%s,%s,%s,%s);"
    query_parameters = (id, name, released_date, rating, ratings_count, downloads, genres, age_rating)
    cursor.execute(query_template, query_parameters)
    #connection.commit() 

3498 Grand Theft Auto V 2013-09-17 4.47 6012 18278 Action Mature
Inserting platforms 1
3328 The Witcher 3: Wild Hunt 2015-05-18 4.67 5653 17194 Action Mature
Inserting platforms 2
4200 Portal 2 2011-04-18 4.62 4984 16180 Shooter Everyone 10+
Inserting platforms 3
5286 Tomb Raider (2013) 2013-03-05 4.05 3498 14287 Action Mature
Inserting platforms 4
4291 Counter-Strike: Global Offensive 2012-08-21 3.57 3050 13939 Action Mature
Inserting platforms 5
5679 The Elder Scrolls V: Skyrim 2011-11-11 4.42 4123 13841 Action Mature
Inserting platforms 6
12020 Left 4 Dead 2 2009-11-17 4.09 2910 13727 Action Mature
Inserting platforms 7
13536 Portal 2007-10-09 4.51 4152 13689 Adventure Teen
Inserting platforms 8
4062 BioShock Infinite 2013-03-26 4.39 3630 13308 Action Mature
Inserting platforms 9
802 Borderlands 2 2012-09-18 4.02 2915 13143 Action Mature
Inserting platforms 10
3439 Life is Strange 2015-01-29 4.11 3289 13131 Adventure Mature
Inserting platforms 11
28 Red Dead Redemption 2 2018-10-26 

In [ ]:
sql = 'select * from platforms;'
cursor.execute(sql)
result = cursor.fetchall()
frame = pd.DataFrame(result)
frame

pid               name
0     1            Neo Geo
1     2                 PC
2     3            Android
3     4           Nintendo
4     5                Web
5     6    Apple Macintosh
6     7  Commodore / Amiga
7     8                iOS
8     9               Xbox
9    10        PlayStation
10   11               SEGA
11   12              Linux

In [ ]:
# sql = ' drop TABLE devices;'
# cursor.execute(sql)

In [ ]:
#mapping the platform table index wise
platform_map = {}
for index, element in enumerate(platform_set):
  platform_map[element] = index+1
print(platform_map)

{'Neo Geo': 1, 'PC': 2, 'Android': 3, 'Nintendo': 4, 'Web': 5, 'Apple Macintosh': 6, 'Commodore / Amiga': 7, 'iOS': 8, 'Xbox': 9, 'PlayStation': 10, 'SEGA': 11, 'Linux': 12}


In [ ]:
#insert into the table 'devices'
for index, row in df1.iterrows():
  id = row['id']
  platforms_list = row['platforms']
  #print()
  for element in platforms_list:
    pid = int(platform_map[element])
    #print(element+" "+str(pid))
    print(pid, id)
    print("Inserting platforms " + str(index+1))
    query_template ="INSERT INTO devices (id, pid) VALUES(%s, %s)"
    query_parameters = (id, pid)
    cursor.execute(query_template, query_parameters)

Streaming output truncated to the last 5000 lines.
2 4179
Inserting platforms 299
10 4179
Inserting platforms 299
9 4179
Inserting platforms 299
4 4179
Inserting platforms 299
2 50778
Inserting platforms 300
10 50778
Inserting platforms 300
9 50778
Inserting platforms 300
2 14016
Inserting platforms 301
9 14016
Inserting platforms 301
6 14016
Inserting platforms 301
2 5161
Inserting platforms 302
10 5161
Inserting platforms 302
9 5161
Inserting platforms 302
8 5161
Inserting platforms 302
6 5161
Inserting platforms 302
12 5161
Inserting platforms 302
2 9793
Inserting platforms 303
6 9793
Inserting platforms 303
12 9793
Inserting platforms 303
2 2047
Inserting platforms 304
10 2047
Inserting platforms 304
9 2047
Inserting platforms 304
8 2047
Inserting platforms 304
3 2047
Inserting platforms 304
2 2597
Inserting platforms 305
10 2597
Inserting platforms 305
9 2597
Inserting platforms 305
8 2597
Inserting platforms 305
3 2597
Inserting platforms 305
4 2597
Inserting platforms 305
2 2166

#d4: Data analysis

**Question 1: What are the top 10 popular game genres?**

In [ ]:
#to get top 10 genres and their count of games
top_10_genres = df1['genres'].value_counts(ascending=True).tail(10)
print(top_10_genres)

Arcade          5
Indie           7
Sports         12
Shooter        15
Simulation     15
Racing         21
RPG            50
Strategy       61
Adventure      97
Action        709
Name: genres, dtype: int64


In [ ]:
#visualization of top 10 genres and their count of games
fig =px.bar(top_10_genres, title='What are the top 10 popular game genres?', text='value',orientation='h',
       labels={
           'value': "", #removing titles for x and y axes
           'index': ""
       })
fig.update_layout(showlegend=False)
fig.show()

Insights:
* Action genre is the most popular genre. 
* 709 games out of 1000 belong to this genre. The 2nd popular is the Adventure genre.


**Question 2: How many games are releasd through years?**

In [ ]:
#to see the columns in the dataframe df1
df1.columns

Index(['id', 'name', 'released_date', 'rating', 'ratings_count', 'downloads',
       'platforms', 'genres', 'age_rating'],
      dtype='object')

In [ ]:
#insert a new column 'released year' in the dataframe
df1.insert(3,"released_year", "")

In [ ]:
#splitting the year from the dates in the column 'released_date' and putting it in the column 'released_year'
for index, row in df1.iterrows():
  df1['released_year'].iloc[index] = row['released_date'].split('-')[0]
df1

/usr/local/lib/python3.8/dist-packages/pandas/core/indexing.py:1732: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



id                              name released_date released_year  \
0      3498                Grand Theft Auto V    2013-09-17          2013   
1      3328          The Witcher 3: Wild Hunt    2015-05-18          2015   
2      4200                          Portal 2    2011-04-18          2011   
3      5286                Tomb Raider (2013)    2013-03-05          2013   
4      4291  Counter-Strike: Global Offensive    2012-08-21          2012   
..      ...                               ...           ...           ...   
995   52330                       The Council    2018-03-12          2018   
996  455532              Star Wars: Squadrons    2020-10-02          2020   
997   28627                            Halo 4    2012-11-06          2012   
998     703                        LET IT DIE    2016-12-03          2016   
999   11407                          Banished    2014-02-17          2014   

     rating  ratings_count  downloads  \
0      4.47           6012      18278   
1      4.67           5653      17194   
2      4.62           4984      16180   
3      4.05           3498      14287   
4      3.57           3050      13939   
..      ...            ...        ...   
995    3.63            287       2043   
996    3.23            204       2039   
997    3.80            520       2038   
998    2.97            313       2038   
999    3.85            276       2037   

                                           platforms    genres    age_rating  
0                            [PC, PlayStation, Xbox]    Action        Mature  
1                  [PC, PlayStation, Xbox, Nintendo]    Action        Mature  
2    [PC, PlayStation, Xbox, Apple Macintosh, Linux]   Shooter  Everyone 10+  
3           [PC, PlayStation, Xbox, Apple Macintosh]    Action        Mature  
4                            [PC, PlayStation, Xbox]    Action        Mature  
..                                               ...       ...           ...  
995                          [PC, PlayStation, Xbox]    Action        Mature  
996                          [PC, PlayStation, Xbox]    Action          Teen  
997                                       [PC, Xbox]    Action        Mature  
998                                [PC, PlayStation]    Action        Mature  
999                                             [PC]  Strategy          Teen  

[1000 rows x 10 columns]

In [ ]:
#created a pivot table to get the release_year and the count of games for each year
pivot = df1.pivot_table(
         index='released_year',
         aggfunc = 'size'
)
pivot

released_year
1993      1
1994      4
1995      2
1996      4
1997      9
1998      4
1999      6
2000      8
2001     11
2002      9
2003     10
2004     17
2005     14
2006     15
2007     22
2008     31
2009     41
2010     53
2011     56
2012     80
2013     89
2014     86
2015     96
2016    102
2017     71
2018     66
2019     42
2020     33
2021     11
2022      7
dtype: int64

In [ ]:
#converted pivot table to a dataframe
df2 = pd.DataFrame(pivot)


In [ ]:
#visualization of the number of games through years
fig = px.line(pivot,
                 title="How many games are releasd through years?",
                 labels={'x':'Years', 'y':'Number of Games'}
                )
fig.update_layout(
    title="How many games are releasd through years?",
    xaxis_title="",
    yaxis_title="Number of Games",
    )
fig.update_layout(showlegend=False)
fig.show()

Insights:
* The highest number of games released in the year 2016. 
* Looking at the line chart, we can say that period between 2012 - 2017 has the highest number of releases.

**Question 3: What are the top 8 Gaming Platforms with highest number of game releases?**

In [ ]:
#sql query to get number of games in different platforms using group by on the 3 tables 'games', 'platforms', 'devices'
sql = 'select p.name, count(p.pid) as \'count_games\' from games g, platforms p, devices d where g.id = d.id and d.pid = p.pid group by d.pid;'
cursor.execute(sql)
result = cursor.fetchall()

df = pd.DataFrame(result)
df

name  count_games
0             Neo Geo            1
1                  PC          961
2             Android          128
3            Nintendo          324
4                 Web            8
5     Apple Macintosh          405
6   Commodore / Amiga            4
7                 iOS          166
8                Xbox          683
9         PlayStation          712
10               SEGA            9
11              Linux          286

In [ ]:
#Visualization of number of games in different platforms
df = df.sort_values(by='count_games', ascending = False).head(8)
fig = px.bar(df, x='name', y='count_games',title='What are the top 8 Gaming Platforms with highest number of game releases?',text='count_games',
             labels={
           'count_games': "",#removing titles for x and y axis
           'name': ""
       }
      )
fig.show()

Insights:
* The PC platform has the highest number of games. 
* The Playstation and Xbox stand in 2nd and 3rd places. 
* Looking at the bar chart, we can say that there are least number of games in the mobile phones like IOS and Android.

**Question 4: What are the top 10 rated games?**

In [ ]:
#to see the dataframe
df1.head(10)

id                              name released_date released_year  \
0   3498                Grand Theft Auto V    2013-09-17          2013   
1   3328          The Witcher 3: Wild Hunt    2015-05-18          2015   
2   4200                          Portal 2    2011-04-18          2011   
3   5286                Tomb Raider (2013)    2013-03-05          2013   
4   4291  Counter-Strike: Global Offensive    2012-08-21          2012   
5   5679       The Elder Scrolls V: Skyrim    2011-11-11          2011   
6  12020                     Left 4 Dead 2    2009-11-17          2009   
7  13536                            Portal    2007-10-09          2007   
8   4062                 BioShock Infinite    2013-03-26          2013   
9    802                     Borderlands 2    2012-09-18          2012   

   rating  ratings_count  downloads  \
0    4.47           6012      18278   
1    4.67           5653      17194   
2    4.62           4984      16180   
3    4.05           3498      14287   
4    3.57           3050      13939   
5    4.42           4123      13841   
6    4.09           2910      13727   
7    4.51           4152      13689   
8    4.39           3630      13308   
9    4.02           2915      13143   

                                           platforms     genres    age_rating  
0                            [PC, PlayStation, Xbox]     Action        Mature  
1                  [PC, PlayStation, Xbox, Nintendo]     Action        Mature  
2    [PC, PlayStation, Xbox, Apple Macintosh, Linux]    Shooter  Everyone 10+  
3           [PC, PlayStation, Xbox, Apple Macintosh]     Action        Mature  
4                            [PC, PlayStation, Xbox]     Action        Mature  
5                  [PC, PlayStation, Xbox, Nintendo]     Action        Mature  
6                 [PC, Xbox, Apple Macintosh, Linux]     Action        Mature  
7  [PC, PlayStation, Xbox, Android, Apple Macinto...  Adventure          Teen  
8           [PC, PlayStation, Xbox, Linux, Nintendo]     Action        Mature  
9  [PC, PlayStation, Xbox, Android, Apple Macinto...     Action        Mature

In [ ]:
#created an sql query to get the top rated games and their names from the table 'games'
sql = 'select name, rating from games g order by rating desc limit 10;'
cursor.execute(sql)
resultDfr = cursor.fetchall()
result_dfr = pd.DataFrame(resultDfr)
result_dfr = result_dfr.sort_values(by = 'rating', ascending = True)
result_dfr

#same using a new dataframe instead of an sql query
# top_10_rated_shows = df1.sort_values(by='rating', ascending = True).tail(10)
# top_10_rated_shows = top_10_rated_shows[['name','rating']]
# top_10_rated_shows

name  rating
8  The Legend of Zelda: Breath of the Wild    4.56
9                             It Takes Two    4.56
7           Mass Effect: Legendary Edition    4.57
6                           The Last Of Us    4.58
4                    Red Dead Redemption 2    4.59
5                               God of War    4.59
3                                 Portal 2    4.62
2                 The Witcher 3: Wild Hunt    4.67
1                The Last Of Us Remastered    4.69
0                          Persona 5 Royal    4.78

In [ ]:
#visualization for the top 10 rated games
fig =px.bar(result_dfr, x='rating', y='name', title='What are the top 10 rated games?', text='rating',
       labels={
           'name': "", #removing titles for x and y axes
           'rating': "Rating"
       })
fig.update_layout(showlegend=False)
fig.show()

Insight: 
* The highes rated game is 'Persona 5 Royal' with 4.78 rating.

**Question 5: What are the top 10 downloaded games?**

In [ ]:
#sql query to get top 10 downloaded games
# sql = 'select name, downloads from games order by downloads desc limit 10;'
# cursor.execute(sql)
# download_dfr = cursor.fetchall()
# download_df = pd.DataFrame(download_dfr)

#to get top 10 downloaded games from the dataframe
top_10_downloads = df1.sort_values(by='downloads', ascending = True).tail(10)
top_10_downloads = top_10_downloads[['name','downloads']]
top_10_downloads

name  downloads
9                     Borderlands 2      13143
8                 BioShock Infinite      13308
7                            Portal      13689
6                     Left 4 Dead 2      13727
5       The Elder Scrolls V: Skyrim      13841
4  Counter-Strike: Global Offensive      13939
3                Tomb Raider (2013)      14287
2                          Portal 2      16180
1          The Witcher 3: Wild Hunt      17194
0                Grand Theft Auto V      18278

In [ ]:
#visualization for top 10 downloaded games
fig =px.bar(top_10_downloads, x='downloads', y='name', title='What are the top 10 downloaded games?', text='downloads',orientation='h',
       labels={
           'downloads': "Number of Downloads", #removing titles for x and y axes
           'name': ""
       }
      )
fig.update_layout(showlegend=False)
fig.show()

Insights:
* The game 'Grand Theft Auto V' is the winner with highest number of downloads so far.
* The 2nd place goes to 'The Witcher 3: Wild Hunt'

#d5: A summary of the key findings

**What are the top 10 popular game genres?**

* Action genre is the most popular genre. 
* 709 games out of 1000 belong to this genre. The 2nd popular is the Adventure genre.

**How many games are releasd through years?**

* The highest number of games released in the year 2016. 
* Looking at the line chart, we can say that period between 2012 - 2017 has the highest number of releases.

**What are the top 8 Gaming Platforms with highest number of game releases?**

* The PC platform has the highest number of games. 
* The Playstation and Xbox stand in 2nd and 3rd places. 
* Looking at the bar chart, we can say that there are least number of games in the mobile phones like IOS and Android.

**What are the top 10 rated games?**

* The highes rated game is 'Persona 5 Royal' with 4.78 rating.

**What are the top 10 downloaded games?**

* The game 'Grand Theft Auto V' is the winner with highest number of downloads so far.
* The 2nd place goes to 'The Witcher 3: Wild Hunt' 

#Future Research

* There are total of 500,000 records in this api endpoint.
* I would like to analyze all the 500,000 records in the dataset if given more time. 
* I would like to see which age categories have the majority of the games?
* I would also analyse the reviews column from the dataset. I would like to identify the keywords from the reviews and analyze how the reviews are for each genre of games.
* I would also analyze the data based on the 'tags' column
* If given more time, I would explore more different visualizations like scatter plot, box plot etc.
